### DISCUSSION TITLE
[ 題目看嘸 ] LeetCode 457. Circular Array Loop

### DISCUSSION THREAD

**Discussion by LPenny-github** [EngTW/English-for-Programmers#126] [Q&A] [answered]
LeetCode:
https://leetcode.com/problems/circular-array-loop/

---

題目敘述：
> You are given a circular array nums of positive and negative integers. If a number k at an index is positive, then move forward k steps. Conversely, if it's negative (-k), move backward k steps. Since the array is circular, you may assume that the last element's next element is the first element, and the first element's previous element is the last element.
>
> Determine if there is a loop (or a cycle) in nums. A cycle must start and end at the same index and the cycle's length > 1. Furthermore, movements in a cycle must all follow a single direction. In other words, a cycle must not consist of both forward and backward movements.

我不能理解 [ 3, 1, 2 ] 為 true，但 
[ -1, -2, -3, -4, -5 ] 為 false 
[ -5, -1, -2, -3, -4 ] 為 false 

如同其他網友的發問：

> In the {3, 1, 2} case we jump and return to the same position. So there is really one step and it is considered a loop. Is it because we "fly over 2 positions"? Fine. Then why {-1, -2, -3, -4, -5} is not a loop? We make one step from -1 to -5 and then we do exactly the same as in case {3, 1, 2) but to the left.

https://leetcode.com/problems/circular-array-loop/discuss/94152/Why-3-1-2-is-a-loop-and-1-2-3-4-5-is-not

題目看不懂，不知道是英文不好，或是漏掉什麼 orz。

麻煩大大解惑 orz 感激不盡

**Comment by twy30** [comment, marked as answer]
@LPenny-github 

你好 😊

> 我不能理解 [ 3, 1, 2 ] 為 true，但
> [ -1, -2, -3, -4, -5 ] 為 false
> [ -5, -1, -2, -3, -4 ] 為 false

就以上你列出的 3 組輸入資料而言，可以由題目中的這個條件來判定：

> > A cycle must start and end at the same index and the cycle's length > 1.

* 迴圈的起點、終點必須相同。
* 迴圈的長度大於 1 。

---

`[ 3, 1, 2 ]` 有一個滿足題目條件的迴圈：

* `nums[1] == 1` → `nums[2] == 2` → `nums[1] == 1`; 從 `nums[1]` 出發、回到 `nums[1]`, 迴圈長度為 2, 有大於 1 。

---

`[ -1, -2, -3, -4, -5 ]` 沒有任何滿足題目條件的迴圈，因為：

* `nums[0] == -1` → `nums[4] == -5` → `nums[4] == -5`; 會一直卡在 `nums[4]`, 無法回到起點的 `nums[0]` 。
* `nums[1] == -2` → `nums[4] == -5` → `nums[4] == -5`; 會一直卡在 `nums[4]`, 無法回到起點的 `nums[1]` 。
* `nums[2] == -3` → `nums[4] == -5` → `nums[4] == -5`; 會一直卡在 `nums[4]`, 無法回到起點的 `nums[2]` 。
* `nums[3] == -4` → `nums[4] == -5` → `nums[4] == -5`; 會一直卡在 `nums[4]`, 無法回到起點的 `nums[3]` 。
* `nums[4] == -5` → `nums[4] == -5`; 雖然從 `nums[4]` 出發、回到了 `nums[4]`, 但迴圈長度為 1, 沒有大於 1 。

---

`[ -5, -1, -2, -3, -4 ]` 的情形與 `[ -1, -2, -3, -4, -5 ]` 相似。

**Reply by LPenny-github** [reply]
感謝 @twy30 大大 orz

我想就是：

1.  有任何滿足題目條件的迴圈，即成立
2.  有任何不滿足題目條件的迴圈，即不成立

我選 2，但題目指的是 1。

**Reply by twy30** [reply]
是的，題目開的要求是這個：

> > Determine if there is a loop (or a cycle) in nums.

「確認 是否 有 一個 迴圈 在`nums`中 。」

* `[ 3, 1, 2 ]` 有 2 個長度大於 1 的迴圈。
* `[ -1, -2, -3, -4, -5 ]` 有 0 個長度大於 1 的迴圈。
* `[ -5, -1, -2, -3, -4 ]` 有 0 個長度大於 1 的迴圈。

---

若更仔細地檢視這句話：

> > Determine if there is a loop (or a cycle) in nums.

從它的語意判斷，它可以寫得更嚴謹：

"Determine if there is **at least** one loop (or a cycle) in nums."

「確認 是否 有 至少 一個 迴圈 在`nums`中 。」

**Reply by LPenny-github** [reply]
@twy30  十分感謝 orz
